internal/handlers/common/projection.go

```diff
@@ -133,102 +171,209 @@ func projectDocument(inclusion bool, doc *types.Document, projection *types.Docu
-	for _, projectionType := range projectionVal.Keys() {
-		supportedProjections := []string{"$elemMatch"}
-		if !slices.Contains(supportedProjections, projectionType) {
-			return fmt.Errorf("projecion %s is not supported", projectionType)
+func applyDocProjection(k1 string, doc *types.Document, k1Projection *types.Document) (err error) {
+	for _, projectionName := range k1Projection.Keys() {
+		if projectionName != "$elemMatch" {
+			panic(projectionName + " not supported!") // checks must be done in projection check func above
 		}
+		var found int
+		conditions := must.NotFail(k1Projection.Get(projectionName)).(*types.Document)
+		found, err = findDocElemMatch(k1, doc, conditions)
+		if err != nil {
+			return
+		}
+		if found < 0 {
+			doc.Remove(k1)
+			return
+		}
+	}
+	return
+}
 
-		// for now it's only $elemMatch further
-		// if the corresponding value is not an array, skip
+func findInArray(k1, k2 string, value any, doc *types.Document, compareRes []types.CompareResult) (found int, err error) {
+	docValueArray := must.NotFail(doc.GetByPath(k1)).(*types.Array)
 
-		var docValueA any
-		docValueA, err = doc.GetByPath(k1)
+	found = -1
+	for j := 0; j < docValueArray.Len(); j++ {
+		e, err := docValueArray.Get(j)
 		if err != nil {
 			continue
 		}
 
-		// $elemMatch works only for arrays, it must be an array
-		docValueArray, ok := docValueA.(*types.Array)
-		if !ok {
-			doc.Remove(k1)
-			return
+		if found >= 0 {
+			doc.RemoveByPath(k1, strconv.Itoa(j))
+			j -= 1
+			continue
 		}
+		switch e := e.(type) {
+		case *types.Document:
+			var d any
+			d, err = e.Get(k2)
+			if err != nil {
+				doc.RemoveByPath(k1, strconv.Itoa(j))
+				j -= 1
+				continue
+			}
+			switch value := value.(type) {
+			case *types.Document, *types.Array: // TODO
+			default:
+				cmp := types.Compare(d, value)
+				if slices.Contains(compareRes, cmp) {
+					found = j
+					continue
+				}
+			}
 
-		// get the elemMatch conditions
-		conditions := must.NotFail(projectionVal.Get(projectionType)).(*types.Document)
+			doc.RemoveByPath(k1, strconv.Itoa(j))
+			j -= 1
 
-		var found int
-		found, err = filterFieldArrayElemMatch(k1, doc, conditions, docValueArray)
-		if found < 0 {
-			doc.Remove(k1)
-			return
+		default:
+			doc.RemoveByPath(k1, strconv.Itoa(j))
+			j -= 1
+			continue
 		}
 	}
+
+	if found < 0 {
+		doc.RemoveByPath(k1)
+		return
+	}
 	return
 }
 
-// filterFieldArrayElemMatch is for elemMatch conditions.
-func filterFieldArrayElemMatch(k1 string, doc, conditions *types.Document, docValueArray *types.Array) (found int, err error) {
-	for k2ConditionField, conditionValue := range conditions.Map() {
-		switch elemMatchFieldCondition := conditionValue.(type) {
-		case *types.Document: // TODO field2: { $gte: 10 }
+// findDocElemMatch is for elemMatch conditions.
+func findDocElemMatch(k1 string, doc, conditions *types.Document) (found int, err error) {
+	found = -1 // >= 0 means found
```

Comment:
Can we use `bool` instead of `int` here?

Reply:
no, I store here the index to remove array item later,

Reply:
changed where possible

Reply:
> no, I store here the index to remove array item later,

This is something that would be done later?

Reply:
@w84thesun it's removed here and left only in `findInArray` func.